plugins/genesys/fu-genesys-usbhub-firmware.c

```diff
@@ -336,50 +303,33 @@ fu_genesys_usbhub_firmware_parse(FuFirmware *firmware,
 	code_size = fu_firmware_get_size(firmware);
 
 	/* calculate checksum */
-	if ((flags & FWUPD_INSTALL_FLAG_IGNORE_CHECKSUM) == 0)
+	if ((flags & FWUPD_INSTALL_FLAG_IGNORE_CHECKSUM) == 0) {
 		if (!fu_genesys_usbhub_firmware_verify(fw, offset, code_size, error))
 			return FALSE;
+	}
 
 	/* get firmware version */
 	if (!fu_genesys_usbhub_firmware_query_version(firmware, fw, offset, error))
 		return FALSE;
 
 	/* parse remaining firmware bytes */
-	if (bufsz > offset + code_size) {
-		gsize fw_offset = offset + code_size;
-
-		do {
-			FuGenesysFwType fw_type =
-			    fu_genesys_usbhub_firmware_query_magic(fw, fw_offset, error);
-			g_autoptr(FuFirmware) firmware_sub = NULL;
-
-			switch (fw_type) {
-			case FU_GENESYS_FW_TYPE_HUB:
-				g_set_error_literal(error,
-						    FWUPD_ERROR,
-						    FWUPD_ERROR_NOT_SUPPORTED,
-						    "fw bytes have dual hub firmware");
-				return FALSE;
-			case FU_GENESYS_FW_TYPE_DEV_BRIDGE:
-				firmware_sub = fu_genesys_usbhub_dev_firmware_new();
-				break;
-			case FU_GENESYS_FW_TYPE_PD:
-				firmware_sub = fu_genesys_usbhub_pd_firmware_new();
-				break;
-			case FU_GENESYS_FW_TYPE_CODESIGN:
-				firmware_sub = fu_genesys_usbhub_codesign_firmware_new();
-				break;
-			default:
-				return FALSE;
-			}
-			/* parse firmware */
-			if (!fu_firmware_parse_full(firmware_sub, fw, fw_offset, flags, error))
-				return FALSE;
-			fu_firmware_add_image(firmware, firmware_sub);
-			/* shift offset */
-			fw_offset = fu_firmware_get_offset(firmware_sub) +
-				    fu_firmware_get_size(firmware_sub);
-		} while (bufsz > fw_offset);
+	offset += code_size;
+	while (offset < bufsz) {
+		g_autoptr(FuFirmware) firmware_sub = NULL;
+		firmware_sub = fu_firmware_new_from_gtypes(fw,
+							   offset,
+							   flags | FWUPD_INSTALL_FLAG_NO_SEARCH,
+							   error,
+							   FU_TYPE_GENESYS_USBHUB_DEV_FIRMWARE,
+							   FU_TYPE_GENESYS_USBHUB_PD_FIRMWARE,
+							   FU_TYPE_GENESYS_USBHUB_CODESIGN_FIRMWARE,
+							   G_TYPE_INVALID);
+		if (firmware_sub == NULL) {
+			g_prefix_error(error, "fw bytes have dual hub firmware: ");
+			return FALSE;
+		}
+		fu_firmware_add_image(firmware, firmware_sub);
```

Comment:
The offset of `FuFirmware` is gone with `FWUPD_INSTALL_FLAG_NO_SEARCH`. Should add `fu_firmware_set_offset(firmware_sub, offset);` here.

Reply:
Good catch, fixed!